shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -404,91 +415,247 @@ public void run() {
     }
   }
 
+  // Keep a queue of ImageReaders.
+  // Each ImageReader holds acquired Images.
+  // When we acquire the next image, close any ImageReaders that don't have any
+  // more pending images.
   @Keep
   @TargetApi(29)
   final class ImageReaderSurfaceProducer
       implements TextureRegistry.SurfaceProducer, TextureRegistry.ImageConsumer {
     private static final String TAG = "ImageReaderSurfaceProducer";
     private static final int MAX_IMAGES = 4;
 
+    private static final boolean VERBOSE_LOGS = false;
+
     private final long id;
 
     private boolean released;
     private boolean ignoringFence = false;
 
     private int requestedWidth = 0;
     private int requestedHeight = 0;
+    private boolean createNewReader = true;
+
+    private long lastDequeueTime = 0;
+    private long lastQueueTime = 0;
+    private long lastScheduleTime = 0;
+
+    private Object lock = new Object();
```

Comment:
Partially for my own knowledge:

1. Is it just worth using thread-safe collections?
2. If not, maybe file a bug that's worth considering _ErrorProne_ (i.e. https://github.com/google/error-prone/blob/master/docs/bugpattern/GuardedBy.md)

Reply:
1. We have multiple collections here that must be kept in sync (the queue and the map) so using thread-safe collections wouldn't help here.
2. I didn't know that GuardedBy exists- we should definitely adopt this, filed: https://github.com/flutter/flutter/issues/142242